compiler/rustc_mir_transform/src/check_niches.rs

```diff
@@ -0,0 +1,488 @@
+        // our current location after every insertion. By iterating backwards, we dodge this issue:
+        // The only Locations that an insertion changes have already been handled.
+        for block in (0..basic_blocks.len()).rev() {
+            let block = block.into();
+            for statement_index in (0..basic_blocks[block].statements.len()).rev() {
+                let location = Location { block, statement_index };
+                let statement = &basic_blocks[block].statements[statement_index];
+                let source_info = statement.source_info;
+
+                let mut finder = NicheFinder { tcx, local_decls, param_env, places: Vec::new() };
+                finder.visit_statement(statement, location);
+
+                for (place, ty, niche) in core::mem::take(&mut finder.places) {
+                    with_no_trimmed_paths!(debug!("Inserting niche check for {:?}", ty));
+                    let (block_data, new_block) = split_block(basic_blocks, location);
+                    finder.insert_niche_check(block_data, new_block, place, niche, source_info);
+                }
+            }
+        }
+    }
+}
+
+struct NicheFinder<'a, 'tcx> {
+    tcx: TyCtxt<'tcx>,
+    local_decls: &'a mut IndexVec<Local, LocalDecl<'tcx>>,
+    param_env: ParamEnv<'tcx>,
+    places: Vec<(Operand<'tcx>, Ty<'tcx>, NicheKind)>,
+}
+
+impl<'a, 'tcx> Visitor<'tcx> for NicheFinder<'a, 'tcx> {
+    fn visit_rvalue(&mut self, rvalue: &Rvalue<'tcx>, location: Location) {
+        if let Rvalue::Cast(CastKind::Transmute, op, ty) = rvalue {
+            if let Some(niche) = self.get_niche(*ty) {
+                self.places.push((op.clone(), *ty, niche));
+            }
+        }
+
+        self.super_rvalue(rvalue, location);
+    }
+
+    fn visit_place(&mut self, place: &Place<'tcx>, context: PlaceContext, location: Location) {
+        match context {
+            PlaceContext::NonMutatingUse(
+                NonMutatingUseContext::Inspect
+                | NonMutatingUseContext::Copy
+                | NonMutatingUseContext::Move,
+            ) => {}
+            _ => {
+                return;
+            }
+        }
+
+        let ty = place.ty(self.local_decls, self.tcx).ty;
+        let Some(niche) = self.get_niche(ty) else {
+            return;
+        };
+
+        if ty.is_ref() {
+            // References are extremely common, checking them imposes a lot of runtime overhead.
+            return;
+        }
+
+        self.places.push((Operand::Copy(*place), ty, niche));
+
+        self.super_place(place, context, location);
+    }
+}
+
+impl<'a, 'tcx> NicheFinder<'a, 'tcx> {
+    fn insert_niche_check(
+        &mut self,
+        block_data: &mut BasicBlockData<'tcx>,
+        new_block: BasicBlock,
+        operand: Operand<'tcx>,
+        niche: NicheKind,
+        source_info: SourceInfo,
+    ) {
+        let type_name =
+            with_no_trimmed_paths!(format!("{:?}", operand.ty(self.local_decls, self.tcx)));
```

Comment:
Is there a reason you are not letting the codegen backend do this? That should reduce the crate metadata size by avoiding a string to be stored.

Reply:
Maybe you should comment at #104862? :p

We found a mis-compilation in this PR. This PR is used to validate the fix.